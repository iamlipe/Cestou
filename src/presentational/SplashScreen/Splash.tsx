import React, {useLayoutEffect} from 'react';
import styled from 'styled-components/native';
import {Dimensions} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {RoutesParamList} from '@/routes';
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
} from 'react-native-reanimated';

import Logo from '@/assets/svgs/logo.svg';
import LogoBasket from '@/assets/svgs/logo-basket.svg';

type NavProps = NativeStackNavigationProp<RoutesParamList, 'AppStack'>;

const {width, height} = Dimensions.get('window');

export const Splash = () => {
  const translateYMontainOne = useSharedValue(width * 3);
  const translateYMontainTwo = useSharedValue(width * 2);
  const scaleLogo = useSharedValue(0);
  const {navigate} = useNavigation<NavProps>();

  const animatedMontainOne = useAnimatedStyle(() => {
    return {
      transform: [
        {translateY: translateYMontainOne.value},
        {translateX: -width * 0.2},
      ],
    };
  });

  const animatedMontainTwo = useAnimatedStyle(() => {
    return {
      transform: [
        {translateY: translateYMontainTwo.value},
        {translateX: width * 0.2},
      ],
    };
  });

  const animatedLogo = useAnimatedStyle(() => {
    return {
      transform: [{scale: scaleLogo.value}],
    };
  });

  useLayoutEffect(() => {
    const TIME_ANIMATION = 750;
    const TIME_ANIMATION_LOGO = 500;

    translateYMontainOne.value = withTiming(width * 2, {
      duration: TIME_ANIMATION,
    });

    setTimeout(
      () => (scaleLogo.value = withTiming(1, {duration: TIME_ANIMATION_LOGO})),
      TIME_ANIMATION,
    );

    translateYMontainTwo.value = withTiming(width, {duration: TIME_ANIMATION});

    setTimeout(
      () => navigate('AppStack'),
      TIME_ANIMATION_LOGO + TIME_ANIMATION,
    );
  }, [navigate, scaleLogo, translateYMontainOne, translateYMontainTwo]);

  return (
    <StyledContainer>
      <StyledContainerLogoAnimated style={[animatedLogo]}>
        <StyledLogoBasket />
        <StyledLogo />
      </StyledContainerLogoAnimated>
      <StyledContainerMontainOneAnimated style={[animatedMontainOne]} />
      <StyledContainerMontainTwoAnimated style={[animatedMontainTwo]} />
    </StyledContainer>
  );
};

const StyledContainer = styled.View`
  flex: 1;
  justify-content: flex-end;
`;

const StyledContainerLogoAnimated = styled(Animated.View)``;

const StyledLogoBasket = styled(LogoBasket)`
  top: ${height * 0.9}px;
  left: ${width * 0.275}px;
  z-index: 1;
`;

const StyledLogo = styled(Logo)`
  top: ${height * 0.9}px;
  left: ${width * 0.275}px;
  z-index: 1;
`;

const StyledContainerMontainOneAnimated = styled(Animated.View)`
  width: ${width * 0.8}px;
  height: ${width * 0.8}px;
  border-radius: ${(width * 0.8) / 2}px;
  background-color: ${({theme}) => theme.colors.PRIMARY_800};
`;

const StyledContainerMontainTwoAnimated = styled(Animated.View)`
  width: ${width * 1.6}px;
  height: ${width * 1.6}px;
  border-radius: ${(width * 1.6) / 2}px;
  background-color: ${({theme}) => theme.colors.PRIMARY_500};
`;
