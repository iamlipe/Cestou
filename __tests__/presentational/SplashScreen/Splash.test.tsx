import React from 'react';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@/config/functions/Testing';

import {Splash} from '@/presentational/SplashScreen/Splash';

describe('Splash', () => {
  test('should screen render correctly', () => {
    renderWithThemeProvider(<Splash />);
  });

  test('should matches snapshot', () => {
    matchSnapshotWithProvider(<Splash />);
  });
});
