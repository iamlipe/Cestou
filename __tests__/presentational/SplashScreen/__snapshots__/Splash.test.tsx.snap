// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splash should matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "flex-end",
      },
    ]
  }
>
  <View
    animatedStyle={
      Object {
        "value": Object {
          "transform": Array [
            Object {
              "scale": 0,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      Object {
        "transform": Array [
          Object {
            "scale": 0,
          },
        ],
      }
    }
  >
    <SvgMock
      style={
        Array [
          Object {
            "left": 206.25000000000003,
            "top": 1200.6000000000001,
            "zIndex": 1,
          },
        ]
      }
    />
    <SvgMock
      style={
        Array [
          Object {
            "left": 206.25000000000003,
            "top": 1200.6000000000001,
            "zIndex": 1,
          },
        ]
      }
    />
  </View>
  <View
    animatedStyle={
      Object {
        "value": Object {
          "transform": Array [
            Object {
              "translateY": 2250,
            },
            Object {
              "translateX": -150,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      Object {
        "backgroundColor": "#00843F",
        "borderRadius": 300,
        "height": 600,
        "transform": Array [
          Object {
            "translateY": 2250,
          },
          Object {
            "translateX": -150,
          },
        ],
        "width": 600,
      }
    }
  />
  <View
    animatedStyle={
      Object {
        "value": Object {
          "transform": Array [
            Object {
              "translateY": 1500,
            },
            Object {
              "translateX": 150,
            },
          ],
        },
      }
    }
    collapsable={false}
    style={
      Object {
        "backgroundColor": "#00B761",
        "borderRadius": 600,
        "height": 1200,
        "transform": Array [
          Object {
            "translateY": 1500,
          },
          Object {
            "translateX": 150,
          },
        ],
        "width": 1200,
      }
    }
  />
</View>
`;
